site/book/03-packages/05-updating-a-package.md

```diff
@@ -65,4 +65,132 @@ Once you have successfully updated the package, commit the changes:
+
+The resource-merge strategy performs a structural comparison of each resource using the
+OpenAPI schema. So rather than performing a text-based merge, kpt leverages the
+common structure of KRM resources.
+
+#### Resource identity
+In order to perform a per-resource merge, kpt needs to be able to match a resource in
+the local package with the same resource in the upstream version of the package. It does
+this matching based on the identity of a resource, which is the combination of group,
+kind, name and namespace. So in our wordpress example, the identity of the`Deployment`
+resource is:
+```
+group: apps
+kind: Deployment
+name: wordpress
+namespace: ""
+```
+Changing the name and/or namespace of a resource is a pretty common way to customize
+a package. In order to make sure this doesn't create problems during merge, kpt will
+automatically adding the `# kpt-merge: <namespace>/<name>` comment on the `metadata` 
+field of every resource when getting or updating a package. An example is the `Deployment`
+resource from the wordpress package:
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata: # kpt-merge: /wordpress
+  name: wordpress
+  labels:
+    app: wordpress
+...
+```
+
+#### Merge rules
+kpt performs a 3-way merge for every resource. This means it will use the resource
+in the local package, the updated resource from upstream, as well as the resource
+at the version where the local and upstream package diverged (i.e.
+common ancestor). When discussing the merge rules in detail, we will be referring to
+the three different sources as local, upstream and origin.
+
+In the discussion, we will be referring to non-associative and associative lists. A
+non-associative list either has elements that are scalars or another list, or it has elements
+that are mappings but without an associative key. An example of this in the kubernetes
+API is the command property on containers:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod
+spec:
+  containers:
+    - name: hello
+      image: busybox
+      command: ['sh', '-c', 'echo "Hello, World!"]
+```
+
+An associative list has elements that are mappings and
+one or more of the fields in the mappings are designated as associative keys. An associative key
+(also sometimes referred to as a merge key) is used to identify the "same" elements in two
+different lists for the purpose of merging them. An example from the kubernetes API
+is the list of containers in a pod which uses the `name` property as the merge key:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod
+spec:
+  containers:
+    - name: web
+      image: nginx
+    - name: sidecar
+      image: log-collector
+```
+
+kpt will primarily look for information about
+any associative keys from the OpenAPI schema, but some fields are also automatically recognized as
+associative keys:
+* `mountPath`
+* `devicePath`
+* `ip`
+* `type`
+* `topologyKey`
+* `name`
+* `containerPort`
+
+
+| Resource level |
+| --- |
```

Comment:
This may be missing the suggestion to use table (Single column table is just a list).

Can you formulate this using 4 columns as follows:

C1: Origin
C2: Upstream
C3: Local
C4: Result

Same for other tables below. Otherwise, I still find this hard to read.

Reply:
I agree we should look at the best way to explain this, but I'm not convinced adding tables is the right approach here. It would effectively mean enumerating all the possible options instead of the explaining the fundamental rules which is encoded in the algorithm. There is considerable number of possible permutations, so the result will be four pretty large tables.

Reply:
I think we have two options:

1- either simplify this section by presenting the same information in a better way (original comment here)
2- remove these details (Specifically the `merge rules` content) from the book as it's disrupts the linear flow of a user reading the book. This is really more of a appendix/reference for merge strategy and is unlike other chapters.

Since 1 doesn't sound feasible, I suggest going with 2. These tables can be moved to reference doc for update and linked from here. WDYT?

Reply:
Yeah, I agree with that. Regardless of how we present this data, it probably gets into more detail than we want in the book. I think we will also run into other topics that are too nitty gritty for the book, but also doesn't fit in the reference docs.

Reply:
I have moved the content to the `update` reference docs. I put it in a separate **Details** section at the end of the doc. I don't want to put too much details early so users have to scroll in order to see the support args and flags.

Another challenge is the way we use the markdown headers. In the reference section, even top-level section use header 3 (`###`), so the available header hierarchy in these pages are very limited. Also, our site style makes header 5 text smaller than the main body, which looks weird. Created a separate issue for this: https://github.com/GoogleContainerTools/kpt/issues/2303